torch/csrc/jit/mobile/model_compatibility.cpp

```diff
@@ -0,0 +1,91 @@
+#include <ATen/core/ivalue.h>
+#include <caffe2/serialize/file_adapter.h>
+#include <caffe2/serialize/inline_container.h>
+#include <torch/csrc/jit/mobile/model_compatibility.h>
+#include <torch/csrc/jit/serialization/import_read.h>
+
+#include <string>
+#include <vector>
+
+namespace torch {
+namespace jit {
+
+using caffe2::serialize::FileAdapter;
+using caffe2::serialize::IStreamAdapter;
+using caffe2::serialize::PyTorchStreamReader;
+using caffe2::serialize::ReadAdapterInterface;
+
+namespace {
+
+c10::IValue readArchive(
+    const std::string& archive_name,
+    PyTorchStreamReader& stream_reader) {
+  c10::optional<at::Device> device;
+
+  auto ivalues = torch::jit::readArchiveAndTensors(
+      archive_name, nullptr, nullptr, device, stream_reader);
+  return ivalues;
+}
+
+bool check_zip_file(std::shared_ptr<ReadAdapterInterface> rai) {
+  std::array<uint8_t, 2> first_short{};
+  static constexpr uint8_t first_slot = 0x80;
+  static constexpr uint8_t second_slot = 0x02;
+  rai->read(
+      /*pos=*/0,
+      /*buf=*/&first_short,
+      /*n=*/2,
+      /*what=*/"checking archive");
+  if (first_short[0] == first_slot && first_short[1] == second_slot) {
+    // NB: zip files by spec can start with any data, so technically they might
+    // start with 0x80 0x02, but in practice zip files start with a file entry
+    // which begins with 0x04034b50. Furthermore, PyTorch will never produce zip
+    // files that do not start with the file entry, so it is relatively safe to
+    // perform this check.
+    TORCH_WARN("The zip file might be problematic. Please check it again.");
+    return true;
```

Comment:
This is a little strange. You want to return false when zip file is not ok.

Also how id 0x80 and 0x02 is derived as values.